lib/path/default.nix

```diff
@@ -137,6 +141,84 @@ in /* No rec! Add dependencies on this file at the top. */ {
     subpathInvalidReason value == null;
 
 
+  /* Join subpath strings together using `/`, returning a normalised subpath string.
+
+    Like `concatStringsSep "/"` but safer, specifically:
+
+    - All elements must be valid subpath strings, see `lib.path.subpath.isValid`
+
+    - The result gets normalised, see `lib.path.subpath.normalise`
+
+    - The edge case of an empty list gets properly handled by returning the neutral subpath `"./."`
+
+    Laws:
+
+    - (Associativity):
+
+          subpath.join [ x (subpath.join [ y z ]) ] == subpath.join [ (subpath.join [ x y ]) z ]
+
+    - (Identity) The neutral "./." value is the neutral identity for normalised paths:
+
+          subpath.join [ ] == "./."
+          subpath.join [ (subpath.normalise p) "./." ] == subpath.normalise p
+          subpath.join [ "./." (subpath.normalise p) ] == subpath.normalise p
+
+    - (Normalised) The result is normalised according to `subpath.normalise`:
+
+          subpath.join ps == subpath.normalise (subpath.join ps)
+
+    - For non-empty lists, the implementation is equivalent to normalising the result of `concatStringsSep "/"`.
+      Note that the above laws can be derived from this one.
+
+          ps != [] -> subpath.join ps == subpath.normalise (concatStringsSep "/" ps)
+
+    Type:
+      subpath.join :: [ String ] -> String
+
+    Example:
+      subpath.join [ "foo" "bar/baz" ]
+      => "./foo/bar/baz"
+
+      # normalise the result
+      subpath.join [ "./foo" "." "bar//./baz/" ]
+      => "./foo/bar/baz"
+
+      # passing an empty list results in the current directory
+      subpath.join [ ]
+      => "./."
+
+      # elements must be valid subpath strings
+      subpath.join [ /foo ]
+      => <error>
+      subpath.join [ "" ]
+      => <error>
+      subpath.join [ "/foo" ]
+      => <error>
+      subpath.join [ "../foo" ]
+      => <error>
+  */
+  subpath.join = paths:
+    assert assertMsg
+      # Fast in case all paths are valid
+      (all (path: subpathInvalidReason path == null) paths)
```

Comment:
We shouldn't use redundant parentheses here, because if anyone doubts their understanding of the language, they can use common sense to infer how the operator precedence works here.
Yet, if someone sees redundant parentheses, they may start to think that their understanding of the language isn't good, even if they're already quite productive in it.

Reply:
I like the `subpathInvalidReason` call a bit more over `isValid` because:
- `subpathInvalidReason` is used again later in this function, so we can't remove the dependency on it anyways, and it's nice to see the correspondence of the two instances.
- Currently the public functions of this library only depend on the internal functions. This is because generally the public functions have some extra error checking which isn't present in the internal ones. While `isValid` doesn't have any such error checking, it's still nice to maintain this property imo

Reply:
> it's nice to see the correspondence of the two instances.

> Currently the public functions of this library only depend on the internal functions. This is because generally the public functions have some extra error checking which isn't present in the internal ones. 

I find these rather weak arguments in themselves, but most importantly they miss my point.
Reusing the function obscures the purpose of the check, and reveals internals that are irrelevant at that point, and distracting. All we care about there is whether the path is valid, not how validity is determined (the error is null).

However internal a given piece of code may seem, it's still primarily for people to read, therefore these little things matter a lot. Every time someone has to complete your intended meaning in their head it's wasting (highly constrained) mental resources on something that is not "actually understanding what you're trying to achieve here".

I know that sounds really nitpicky, but I'm also convinced we've achieved the goal of highly maintainable code when someone as lazy and distracted as myself can read through the code linearly without much effort and still be sure to understand it completely. That may be a highly subjective and to often unsuitable measure, but it's the closest to thing I have to imagine how uninitiated readers will deal with the code.

Reply:
Yeah agree that this sounds better, I'm changing it to `isValid` throughout the code